code/datums/components/wire_panel.dm

```diff
@@ -0,0 +1,284 @@
+/**
+ * Wire Panel Component
+ *
+ * Addding this component to an object will add an interactable wire panel.
+ *
+ * Consumer is responsible for icon updates / functionality changes based on wire control.
+ */
+
+TYPEINFO(/datum/component/wirePanel)
+	initialization_args = list(
+		ARG_INFO("definition", DATA_INPUT_REF, "Datum holding our wire panel defintion"),
+	)
+/datum/component/wirePanel
+	/// Reference to a wire defintion datum, passed in Initialize
+	var/datum/wirePanel/panelDefintion/panel_def
+	/// panel cover status (default closed)
+	var/cover_status = WPANEL_COVER_CLOSED
+	/// wire-indexed list of booleans
+	var/list/cut_wires = list()
+	/// bitmask of which wire controls are active (default all active)
+	var/active_wire_controls = ~0
+
+/datum/component/wirePanel/Initialize(_definition)
+	..()
+	if(!istype(parent, /obj))
+		return COMPONENT_INCOMPATIBLE
+	src.panel_def = _definition
+	for(var/datum/wirePanel/indicatorDefintion/indicator in src.panel_def.indicator_lights)
+		indicator.cache_color()
+	for (var/wire in 1 to length(src.panel_def.wire_definitions))
+		src.panel_def.wire_definitions[wire].cache_color()
+		src.cut_wires += list(FALSE) // no wires cut by default
+	RegisterSignal(parent, COMSIG_ATTACKBY, .proc/attackby)
+	RegisterSignal(parent, COMSIG_ATTACKHAND, .proc/attack_hand)
+
+	RegisterSignal(parent, COMSIG_WPANEL_MOB_WIRE_ACT, .proc/mob_wire_act)
+	RegisterSignal(parent, COMSIG_WPANEL_ION_STORM, .proc/ion_storm)
+	RegisterSignal(parent, COMSIG_WPANEL_DISABLE_RANDOM_WIRE, .proc/disable_random_wire)
+
+	RegisterSignal(parent, COMSIG_WPANEL_SET_CONTROL, .proc/set_control)
+	RegisterSignal(parent, COMSIG_WPANEL_SET_COVER, .proc/set_cover)
+
+	RegisterSignal(parent, COMSIG_WPANEL_STATE_CONTROLS, .proc/state_controls)
+	RegisterSignal(parent, COMSIG_WPANEL_STATE_COVER, .proc/state_cover)
+
+	RegisterSignal(parent, COMSIG_WPANEL_UI_DATA, .proc/ui_data)
+	RegisterSignal(parent, COMSIG_WPANEL_UI_STATIC_DATA, .proc/ui_static_data)
+	RegisterSignal(parent, COMSIG_WPANEL_UI_ACT, .proc/ui_act)
+
+/// Handle tool usage on the panel/object
+/datum/component/wirePanel/proc/attackby(obj/parent, obj/item/item, mob/user)
```

Comment:
How would this handle more complex panel opening behaviour, say if you had to swipe an ID before opening the panel?

Reply:
I added `WPANEL_COVER_LOCKED` to cover that general use case; machines would have to add code to change from `WPANEL_COVER_LOCKED` to `WPANEL_COVER_CLOSED` on whatever criteria, like ID swiping.